### ISSUE TITLE
Possibly wrong expectation in fetch/origin/assorted.window.js test

### ISSUE THREAD

**Issue by rwlbuis** [web-platform-tests/wpt#22298]
I am a bit confused about this comment:
// When we're dealing with CORS (mode is "cors"), we shouldn't take the
// Referrer-Policy into account

What about CORS mode fetches that turn out to be same-origin though?

The way I read the spec, first case of 5 from Main fetch applies, setting tainting to basic.
That means Step 3 of https://fetch.spec.whatwg.org/#append-a-request-origin-header
would apply, meaning the referrer-policy would be taken into account.

So it is quite possible I understand the specification wrong but it is also possible (some) test expectations are wrong.

**Comment by rwlbuis** [comment]
@annevk ping

**Comment by youennf** [comment]
"Origin header and POST same-origin fetch cors mode with Referrer-Policy no-referrer" is passing in all browsers and expects an Origin header.

I guess your suggestion would be to update https://fetch.spec.whatwg.org/#append-a-request-origin-header to use request mode instead of request response tainting mode.

**Comment by rwlbuis** [comment]
> "Origin header and POST same-origin fetch cors mode with Referrer-Policy no-referrer" is passing in all browsers and expects an Origin header.
> 
> I guess your suggestion would be to update https://fetch.spec.whatwg.org/#append-a-request-origin-header to use request mode instead of request response tainting mode.

Yes that sounds good. @annevk WDYT?

**Comment by annevk** [comment]
Yeah, I think that's a good idea. I vaguely recall discussing something similar with @JuniorHsu, but I can't find where we discussed that unfortunately.

**Comment by rwlbuis** [comment]
@annevk the bug leading to this discovery landed but it still assumes the old behavior:
https://bugs.webkit.org/show_bug.cgi?id=209066
For the new behavior, what do you propose? Do you want to make a patch for the spec/test change or should I take a a stab?

**Comment by annevk** [comment]
If you could take a stab that'd be great. I think we basically want to include the `Origin` header

* once we've gone across origins at least once for CORS requests
* or if the method is not HEAD/GET (CSRF defense)

**Comment by JuniorHsu** [comment]
> Yeah, I think that's a good idea. I vaguely recall discussing something similar with @JuniorHsu, but I can't find where we discussed that unfortunately.

Please see [here](https://github.com/whatwg/fetch/pull/908#issuecomment-499477191 ) for the previous discussion but I'm happy to see if we have better idea.

**Comment by rwlbuis** [comment]
I created https://github.com/whatwg/fetch/pull/1013 for this.
BTW it does not seem I can push to the fetch repo.
